### ISSUE TITLE
Recursive sequence type of `Sequence[NotAString]` should not accept `str`

### ISSUE THREAD

**Issue by jorenham** [python/mypy#18184]
With at least mypy 1.13.0, the false negative can be reproduced with:

```py
from collections.abc import Sequence
from typing import TypeAlias

class Scalar: ...

Vector: TypeAlias = Sequence[Scalar]
Tensor: TypeAlias = Vector | Sequence["Tensor"]

accepted_vector: Vector = [Scalar()]     # true negative
rejected_vector: Vector = "duck"         # true positive: ducks aren't vectors

accepted_tensor: Tensor = [[[Scalar()]]] # true negative
rejected_tensor1: Tensor = object()      # true positive: objects aren't tensors
rejected_tensor2: Tensor = "duck"        # false negative: ducks are tensors...?

```

[mypy-play](https://mypy-play.net/?mypy=1.13.0&python=3.12&flags=strict&gist=834dea0044d0f8c5d3cd3c5b9ff886f5)

---

originally posted in https://github.com/python/mypy/issues/731#issuecomment-2496165736

**Comment by jorenham** [comment]
> This works as expected (even if surprising) because `str` is a subclass of `Sequence[str]`. 

but `str` is not assignable to `Scalar`

**Comment by jorenham** [comment]
> As an exercise, try writing a code where an actual runtime error happens because of this behavior (the point is that you can't).

```py
>>> import numpy as np
>>> np.array("not a tensor", np.float64)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: could not convert string to float: 'not a tensor'
```

**Comment by jorenham** [comment]
FYI, this works as expected in pyright:

https://pyright-play.net/?strict=true&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGUkBHAVySxiQChRIpUBPCuiwBzMpWq0AKoKQBBeOgI5OnGPGU56MAhpAAuKHmOqAailTVDMivMXKoAXnpNW7JAG06OvQF1OUmw4VlA2dkpazuZo1FAAPi4sbBweAESBuNSp-pwEMBwUqEgAJgD6AG4WIdGWIE5QXj4EIAAUAJS%2BUF1dAMR8IKxQWEgiBBiVnCBIAFYWJRVVBlA1sc6pxcwwANap3Xu9-YNUOOjjSIYb21rNbADktJUxICq5%2BUiF80WZSxnBdc4eQHeXTNdq%2BcFQPqoAZIIYjMboCZTWZoT5BagARms6P%2BUDARBRqHa%2B0hh1hx1OiPOeIJFmuUyw9z4OJUyLmZS%2BfwATNjvvV1psdiSDsBdDhYcNRmcLoL6bDOdR8MYAPycIA

**Comment by davidhalter** [comment]
@ilevkivskyi I also feel like this is a Mypy issue. I am passing AFAIK all of Mypy's relevant recursion tests (except the ones with `--disable-recursive-aliases`) and this case seems to behave correct in my implementation.

**Comment by jorenham** [comment]
@KotlinIsland any thoughts on this?

**Comment by KotlinIsland** [comment]
I think its consistent with the concept/spec, but impractical... maybe. I think pyright is wrong/inconsistent here...

let's change how it works in basedmypy 🚀🚀🚀
